analyzers/src/SonarAnalyzer.CSharp/Rules/SecurityPInvokeMethodShouldNotBeCalled.cs

```diff
@@ -32,64 +29,24 @@ namespace SonarAnalyzer.Rules.CSharp
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     [Rule(DiagnosticId)]
-    public sealed class SecurityPInvokeMethodShouldNotBeCalled : SonarDiagnosticAnalyzer
+    public sealed class SecurityPInvokeMethodShouldNotBeCalled : SecurityPInvokeMethodShouldNotBeCalledBase<SyntaxKind, InvocationExpressionSyntax>
     {
-        internal const string DiagnosticId = "S3884";
-        private const string MessageFormat = "Refactor the code to remove this use of '{0}'.";
-        private const string InteropDllName = "ole32.dll";
+        protected override SyntaxKind SyntaxKind => SyntaxKind.InvocationExpression;
+        protected override ILanguageFacade Language => CSharpFacade.Instance;
 
-        private static readonly ISet<string> InvalidMethods = new HashSet<string>
-        {
-            "CoSetProxyBlanket",
-            "CoInitializeSecurity"
-        };
+        public SecurityPInvokeMethodShouldNotBeCalled() : base(RspecStrings.ResourceManager) { }
 
-        private static readonly DiagnosticDescriptor rule =
-            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        protected override SyntaxNode Expression(InvocationExpressionSyntax invocationExpression) =>
+            invocationExpression.Expression;
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        protected override SyntaxToken Identifier(SyntaxNode syntaxNode) =>
+            ((IdentifierNameSyntax)syntaxNode).Identifier;
```

Comment:
TBH, I didn't see this type cast coming and your original `TIdentifierNameSyntax` was a clearer solution. But with the `SyntaxFacade` on it's way, I guess we can leave it as it is.